### PR TITLE
Allow custom options (WIP)

### PR DESCRIPTION
**Pull request by facelessuser**
Allow custom options and optionally pass them to the formatter.

Ref #349

### CODE REVIEW

**Comment by facelessuser** [comment]
## Todo

- [ ] Need unittests
- [ ] Possibly replace lambda with functools.partial
- [ ] Documentation

File: pymdownx/superfences.py [hunk 1]
```diff
@@ -46,22 +46,18 @@
...
-        (?:[ \t]+(?P<linestep>[\d]+))?                                        #   Line step
-        (?:[ \t]+(?P<linespecial>[\d]+))?                                     #   Line special
-    (?P=quot2)[ \t]*|
-    (?P<tab>tab=)(?:(?P<quot3>"|')(?P<tab_title>.*?)(?P=quot3))?[ \t]*        # Tab specifier
-    )*
-    }?[ \t]*$                                                                 # Language closing
+    (?P<fence>~{3,}|`{3,})[ \t]*                                                          # Fence opening
+    (\{?                                                                                  # Language opening
+    \.?(?P<lang>[\w#.+-]*))?[ \t]*                                                        # Language
+    (?P<options>(?:\b[a-zA-Z][a-zA-Z0-9_]*=(?:(?P<quot>"|')[^"'\n\r]*(?P=quot))?[ \t]*)*) # Options
```
Comment (facelessuser) [inline pymdownx/superfences.py]: This is evaluated line by line, so it is probably not necessary to exclude `\n\r`....not sure what I was thinking here.

**Comment by dbader** [comment]
From an enduser perspective I really like this. Implementing a custom validator seems straightforward and looking at your updated `highlight()` implementation acting on the custom config later on works the way I'd expect. I'm a 👍 This is cool!

**Comment by dbader** [comment]
This might be helpful: https://stackoverflow.com/questions/4764547/creating-dictionary-from-space-separated-key-value-string-in-python, especially the `shlex` example. I think it should be backwards compatible with the current regex-based `hl_lines=...` parser.

File: pymdownx/superfences.py [hunk 2]
```diff
@@ -134,6 +130,34 @@ def fence_div_format(source, language, css_class):
...
+    # Check format of valid keys
+    if okay:
+        for opt, validator in (('hl_lines', RE_HL_LINES), ('linenums', RE_LINENUMS)):
+            if opt in options:
+                value = options[opt]
+                if value is None or validator.match(options[opt]) is None:
+                    okay = False
+                    break
+
+    return okay
```
[the remark this answers is not in the document]
Comment (facelessuser) [reply pymdownx/superfences.py]: Yeah, its a habit of mine to generally have one exit point.  For better or for worse :smile:.  There are times I deviate from this, but it is usually something I do. If it was a long chain of multiple `if okay`, I'd probably just start returning as you suggested.

File: pymdownx/superfences.py [hunk 3]
```diff
@@ -556,19 +606,33 @@ def search_nested(self, lines):
                 lines = lines[:start] + [fenced] + lines[end:]
         return lines
 
-    def highlight(self, src, language):
+    def highlight(self, src, language, options=None):
```
Comment (facelessuser) [inline pymdownx/superfences.py]: It's weird to have options passed in but use `self.options`, I should just use the passed in one.

**Comment by facelessuser** [comment]
`shlex` is indeed a good suggestion, but I generally don't like to pull in libraries when the solution is already quite trivial.  If it cleans up an overly complicated solution, or provides something extra that I feel I need but am missing, than I'm usually on board, but in this particular case I think it may be a little overkill.

**Comment by facelessuser** [comment]
Glad to hear! Hopefully I'll get documentation updated soon and get a released out.

**Pull request merged**